### ISSUE TITLE
Bug: no-magic-numbers has false positive on <N> when ignoreNumericLiteralTypes is true

### ISSUE THREAD

**Issue by ghost** [eslint/eslint#20372]
### Environment

Node version: v24.10.0
npm version: v11.6.0
Local ESLint version: v9.37.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 24.6.0

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```js
"no-magic-numbers": ["error",{"ignoreNumericLiteralTypes": true}]
```
</details>

### What did you expect to happen?

```ts
type bar = 42; // no error
const baz = thing<42>(); // error
```

Neither line should trigger an error for this rule

### Link to Minimal Reproducible Example

[playground link](https://eslint.org/play/#eyJ0ZXh0IjoidHlwZSBiYXIgPSA0MjtcbmNvbnN0IGJheiA9IHRoaW5nPDQyPigpOyIsIm9wdGlvbnMiOnsicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOlsiZXJyb3IiXSwiZm9yLWRpcmVjdGlvbiI6WyJlcnJvciJdLCJnZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOlsiZXJyb3IiXSwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOlsiZXJyb3IiXSwibm8tY2xhc3MtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbXBhcmUtbmVnLXplcm8iOlsiZXJyb3IiXSwibm8tY29uZC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tY29uc3QtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0YW50LWJpbmFyeS1leHByZXNzaW9uIjpbImVycm9yIl0sIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6WyJlcnJvciJdLCJuby1jb250cm9sLXJlZ2V4IjpbImVycm9yIl0sIm5vLWRlYnVnZ2VyIjpbImVycm9yIl0sIm5vLWRlbGV0ZS12YXIiOlsiZXJyb3IiXSwibm8tZHVwZS1hcmdzIjpbImVycm9yIl0sIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6WyJlcnJvciJdLCJuby1kdXBlLWVsc2UtaWYiOlsiZXJyb3IiXSwibm8tZHVwZS1rZXlzIjpbImVycm9yIl0sIm5vLWR1cGxpY2F0ZS1jYXNlIjpbImVycm9yIl0sIm5vLWVtcHR5IjpbImVycm9yIl0sIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1lbXB0eS1wYXR0ZXJuIjpbImVycm9yIl0sIm5vLWVtcHR5LXN0YXRpYy1ibG9jayI6WyJlcnJvciJdLCJuby1leC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjpbImVycm9yIl0sIm5vLWZhbGx0aHJvdWdoIjpbImVycm9yIl0sIm5vLWZ1bmMtYXNzaWduIjpbImVycm9yIl0sIm5vLWdsb2JhbC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW1wb3J0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1pbnZhbGlkLXJlZ2V4cCI6WyJlcnJvciJdLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6WyJlcnJvciJdLCJuby1sb3NzLW9mLXByZWNpc2lvbiI6WyJlcnJvciJdLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1uZXctbmF0aXZlLW5vbmNvbnN0cnVjdG9yIjpbImVycm9yIl0sIm5vLW5vbm9jdGFsLWRlY2ltYWwtZXNjYXBlIjpbImVycm9yIl0sIm5vLW9iai1jYWxscyI6WyJlcnJvciJdLCJuby1vY3RhbCI6WyJlcnJvciJdLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOlsiZXJyb3IiXSwibm8tcmVkZWNsYXJlIjpbImVycm9yIl0sIm5vLXJlZ2V4LXNwYWNlcyI6WyJlcnJvciJdLCJuby1zZWxmLWFzc2lnbiI6WyJlcnJvciJdLCJuby1zZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjpbImVycm9yIl0sIm5vLXNwYXJzZS1hcnJheXMiOlsiZXJyb3IiXSwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOlsiZXJyb3IiXSwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOlsiZXJyb3IiXSwibm8tdW5yZWFjaGFibGUiOlsiZXJyb3IiXSwibm8tdW5zYWZlLWZpbmFsbHkiOlsiZXJyb3IiXSwibm8tdW5zYWZlLW5lZ2F0aW9uIjpbImVycm9yIl0sIm5vLXVuc2FmZS1vcHRpb25hbC1jaGFpbmluZyI6WyJlcnJvciJdLCJuby11bnVzZWQtbGFiZWxzIjpbImVycm9yIl0sIm5vLXVudXNlZC1wcml2YXRlLWNsYXNzLW1lbWJlcnMiOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1iYWNrcmVmZXJlbmNlIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtY2F0Y2giOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1lc2NhcGUiOlsiZXJyb3IiXSwibm8td2l0aCI6WyJlcnJvciJdLCJyZXF1aXJlLXlpZWxkIjpbImVycm9yIl0sInVzZS1pc25hbiI6WyJlcnJvciJdLCJ2YWxpZC10eXBlb2YiOlsiZXJyb3IiXSwibm8tbWFnaWMtbnVtYmVycyI6WyJlcnJvciIseyJpZ25vcmVOdW1lcmljTGl0ZXJhbFR5cGVzIjp0cnVlfV19LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJzb3VyY2VUeXBlIjoibW9kdWxlIn0sInBhcnNlciI6IkB0eXBlc2NyaXB0LWVzbGludC9wYXJzZXIifX19)

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by DMartens** [comment]
Thank you for reporting this but I do not know whether this is a false positive.
The original rule from `typescript-eslint` also reports this ([playground](https://typescript-eslint.io/play/#ts=5.9.2&showAST=es&fileType=.tsx&code=MYewdgzgLgBARgQwF4wLwygCwJZgOYA8ALAEwB8AFAJQDcMA9PTAKYBOrIrQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFsBDAc0plaTWNwBGiaKgwBtSFOgdokADRZsESv07REAOTFShAGUr4pveABUSKDPjiJw2AL4BdF65CugA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)) and we aligned our [implementation](https://github.com/eslint/eslint/blob/3115021439490d1ed12da5804902ebbf8a5e574b/lib/rules/no-magic-numbers.js#L43).

**Comment by ghost** [comment]
So their implementation is broken too.

The name of the rule is "ignoreNumericLiteralTypes".  This is a numeric literal type.  I'm not sure how that can any more clearly be a false positive

**Comment by ghost** [comment]
Opened https://github.com/typescript-eslint/typescript-eslint/issues/11821

**Comment by DMartens** [comment]
This does not mean it is a false positive but I agree the name should include that it should be assigned to a type alias and the documentation for the option should be updated to reflect that.
I do not think `typescript-eslint` wants to update their rule implementation as their rule is frozen as the TypeScript specific code is now in the core rule.

**Comment by ghost** [comment]
Okay, let me try a different approach to convince you.

I do NOT want this rule enabled in TS types.  I don't see any way to accomplish that, even when setting all of the TS specific options.  Regardless of whether you (obviously incorrectly ;)) disagree that the name of this specific param should not ignore this behavior, shouldn't the native eslint, not typescript-eslint, rule have an option at least to ignore TS?

```json
["error",{"ignoreNumericLiteralTypes":true, "ignoreEnums": true, "ignoreReadonlyClassProperties": true, "ignoreTypeIndexes": true}]
```

```
// should error
const A = 42;

// should not error
const B: 42 = bar; // error
type C = 42;
const D = fn<42>()  // error
type E = {x: 42}  // error
```

**Comment by DMartens** [comment]
I think this is correct behavior as the intent of the rule is to disallow using numbers without context.
Can you please provide the motivation for allowing this; especially why this should be allowed when you can assign the number to a type alias (= giving the number context):
```typescript
type Answer = 42;
const value: Answer = 42;
```

**Comment by ghost** [comment]
The correct behavior is to be impossible to disable it on types?!?  🤨 I feel eslint is overreaching here.  This should be typescript-eslint's responsibility.

There is a fundamental difference between numeric literals in JS and in TS types.

The reasoning is simple -- the code you posted is unnecessarily verbose and not necessary in many cases.  We use numeric literals as array lengths, for instance, all the time.

```ts
const foo = expectThings<5>(1, 2, 3, 4); // TS error
```

Changing this to:

```ts
type numExpectedThings = 5;
const foo = expectThings<numExpectedThings>(1, 2, 3, 4); // TS error
```

Is perfectly valid, but the decision of which style we want to use should be ours.

**Comment by kirkwaiblinger** [comment]
@a-non-a-mouse 

Continuing the conversation at https://github.com/typescript-eslint/typescript-eslint/issues/11821 and speaking from the typescript-eslint side...

> 🤨 I feel eslint is overreaching here. This should be typescript-eslint's responsibility.

The boundary between typescript-eslint and ESLint is actively in flux. You can read more about this at https://github.com/eslint/eslint/issues/19173 and https://github.com/eslint/eslint/discussions/18830

In short, no-magic-numbers is a rule that can be fully support TS in ESLint core, so TS support has been added, and they are taking over maintenance of the TS behavior. The typescript-eslint version accordingly is frozen for feature requests, and will likely be deprecated and cease to exist at some point.

> The correct behavior is to be impossible to disable it on types?!?

As mentioned in https://github.com/typescript-eslint/typescript-eslint/issues/11821, the option named `ignoreNumericLiteralTypes` has evidently from the start had the effect of "treat numeric literals in type contexts the same as numeric literals in value contexts; require them to be assigned to a name in order to be used". When the option is not enabled, the rule just flags _all_ numeric literals in type contexts.

My parse of the name "ignoreNumericLiteralTypes" is the same as yours: it would never flag a numeric literal in a type context. And, accordingly, my intuition/opinion is that the current behavior of `ignoreNumericLiteralTypes` should just be the unconditional default behavior of the rule. I suspect but haven't verified that the only reason it's not was to avoid a breaking change when type support was originally written on the typescript-eslint side.

In any case, my current understanding, unless I'm missing something, is that there is not a _bug_ here. There's definitely a case of dubious naming and lacking documentation. And it seems like you have a feature request for an option which exempts any numeric literal types from being flagged by the rule.

**Comment by ghost** [comment]
Don't care that people are too stupid to recognize obvious bugs
